### PR TITLE
Initial support for dynamic imports

### PR DESCRIPTION
**Pull request by ChadKillingsworth**
Allows dynamic import statements to be recognized by the compiler and utilized. Dynamic import expressions can be used as a script or module loader even without full module rewriting. Contains only basic support for type checking in that the type of the import is a Promise resolving to the unknown type.

Left for future PRs:
 - Module rewriting
 - Transpilation (aliases)

See the design document for the full plan: https://docs.google.com/document/d/1thSbGV5i96jltVwn76KYrPkWoNc-yj_KU3HtBGUCgMc/edit#heading=h.c037si364tjn

### CODE REVIEW

File: src/com/google/javascript/jscomp/GatherModuleMetadata.java [hunk 1]
```diff
@@ -233,6 +235,13 @@ private void visitImportOrExport(NodeTraversal t, Node importOrExport) {
       }
     }
 
+    private void visitDynamicImport(NodeTraversal t, Node dynamicImport) {
+      currentModule
+          .metadataBuilder
+          .es6ImportSpecifiersBuilder()
+          .add("*");
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/GatherModuleMetadata.java:242]: Shouldn't this be looking at the children of the dynamic import node?

File: src/com/google/javascript/jscomp/parsing/parser/FeatureSet.java [hunk 2]
```diff
@@ -199,10 +199,8 @@
     // https://github.com/tc39/proposal-optional-catch-binding
     OPTIONAL_CATCH_BINDING("Optional catch binding", LangVersion.ES2019),
 
-    // Stage 3 proposals likely to be part of ES2020
-    DYNAMIC_IMPORT("Dynamic module import", LangVersion.ES_UNSUPPORTED),
-
     // ES 2020 Stage 4
+    DYNAMIC_IMPORT("Dynamic module import", LangVersion.ES2020),
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/parsing/parser/FeatureSet.java:203]: Yes, it's officially in ES_2020, but from our point of view, we don't really support it, so we can't put it in that language version yet. If you move this into our definition of ES_2020, you'll also need to modify the definition of the ES_NEXT FeatureSet to subtract this feature, because we cannot safely output it. (It will most likely cause miscellaneous exceptions in optimization passes that don't expect to see this node type.)

The usual procedure for a feature is something like:

1. It moves to ES_NEXT_IN when we can safely accept it in the input and transpile down to at least ES_NEXT level output (or just fail with an error because we cannot transpile it, since that is "safe" behavior).
2. It moves to ES_NEXT when we can safely carry the feature all the way through compilation and output it.

**Review by brad4d — CHANGES_REQUESTED** [review]

File: src/com/google/javascript/jscomp/GatherModuleMetadata.java (hunk 1 above)
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/GatherModuleMetadata.java:242]: I don't understand the module metadata completely, but these both refer to the same thing:

```js
import * as foo from './foo.js';
import('./foo.js').then((foo2) => {
   console.log(foo === foo2);
});
```

My logic was that a dynamic import was a "*" import so this was simply indicating that the currentModule referred to the module export namespace.

File: src/com/google/javascript/jscomp/parsing/parser/FeatureSet.java (hunk 2 above)
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/parsing/parser/FeatureSet.java:203]: @brad4d Based off the call Thursday, I left dynamic import in the ES_2020 category, but added a warning when any usage is encountered.

**Comment by ChadKillingsworth** [comment]
@brad4d This should be ready for review. I ran my local project with this pass (combined with #3367 and #3724). Dynamic imports were successfully passed all the way through the compiler. I did find a couple of passes that needed small changes to support dynamic imports and those are included in this PR.

**Review by brad4d — APPROVED** [review]

**Comment by brad4d** [comment]
I'll import and test this...

**Comment by brad4d** [comment]
As this is currently written:

1. Dynamic will be rejected unless the language in is specified as `ES_2020` or newer.
2. The compiler will generate a warning (not an error), if the output language is older than `ES_2020`.
3. Regardless of the output language level, the `import(modulePath)` syntax will be left unmodified in the output.

Number 2 above is rather unusual for the compiler. I'd expect either transpilation to occur or an error to be reported.
I believe we're making an exception in this case for these reasons.

1. Many browsers had support for dynamic import long before the other features of `ES_2020`.
2. @ChadKillingsworth currently has a way to post-process the code to replace the `import()` calls with a method that will be available at runtime.
3. @ChadKillingsworth  plans to have a follow-up change that allow the user to specify a function name for the compiler to use to replace the `import`.

**Comment by brad4d** [comment]
@ChadKillingsworth 

It's not going to work to diverge from the usual practice as I described in my last comment.
There's too much logic in the compiler that depends on it having a correct understanding of what features are currently present in the AST. And, that's a good thing, really.

The "transpilation" pass needs to report an error and stop compilation, not a warning.

I think making the dynamic import "work" for `--language_out < ES_2020` should wait for the follow up change that allows specifying a replacement function to call, so the transpilation can actually get rid of the dynamic import syntax.

OR, you could make the transpilation pass change the `DYNAMIC_IMPORT` node into a `CALL` node that just calls a function named `import`. When the code is output, it will look the same.

To make that work, the transpilation will also need to update the externs AST to include a declaration of a global `import` function, so it won't appear to be undeclared. Note that we cannot add such a declaration to our normal externs, because it is a syntax error to declare a function named `import`. Inserting it directly into the externs AST avoids the parsing, so we don't have this probllem.

I've also noticed a few other problems when trying to get this code to work inside Google.
I'll create a new code review for those.

File: src/com/google/javascript/jscomp/AstAnalyzer.java [hunk 3]
```diff
@@ -520,6 +524,7 @@ boolean nodeTypeMayHaveSideEffects(Node n) {
       case FOR_IN: // assigns to a loop LHS
       case FOR_OF: // assigns to a loop LHS, runs an iterator
       case FOR_AWAIT_OF: // assigns to a loop LHS, runs an iterator, async operations.
+      case DYNAMIC_IMPORT:
         return true;
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/AstAnalyzer.java:528]: missing `break;` here

File: src/com/google/javascript/jscomp/DefaultPassConfig.java [hunk 4]
```diff
@@ -261,6 +262,11 @@ private void addModuleRewritingPasses(List<PassFactory> checks, CompilerOptions
       checks.add(rewriteCommonJsModules);
     }
 
+    if (options.getLanguageIn().toFeatureSet().contains(Feature.DYNAMIC_IMPORT)) {
+      // Temporary location until this pass does more than warn about usage
+      checks.add(es2020RewriteDynamicImport);
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/DefaultPassConfig.java:267]: Instead of being added here, the pass should be added in `addModuleRewritingPasses()`.

Also, it should be added by calling a new method `TranspilationPasses.addDynamicImportPass(checks)`.
We want to keep the transpilation pass creation in the `TranspilationPasses` class.

File: src/com/google/javascript/jscomp/Es2020RewriteDynamicImport.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
+package com.google.javascript.jscomp;
+
+import com.google.javascript.rhino.Node;
+
+public class Es2020RewriteDynamicImport implements CompilerPass, NodeTraversal.Callback {
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/Es2020RewriteDynamicImport.java:5]: need a JavaDoc comment for this class

File: src/com/google/javascript/jscomp/DefaultPassConfig.java (hunk 4 above)
Comment (brad4d) [reply src/com/google/javascript/jscomp/DefaultPassConfig.java:267]: See `TranspliationPasses.addEs6ModulePass()` as an example.

File: src/com/google/javascript/jscomp/Es2020RewriteDynamicImport.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+          "JSC_DYNAMIC_IMPORT_EXPERIMENTAL",
+          "Dynamic import support is experimental. No module rewriting is performed.");
+
+  public Es2020RewriteDynamicImport(AbstractCompiler compiler) {
+    this.compiler = compiler;
+  }
+
+  @Override
+  public void process(Node externs, Node root) {
+    NodeTraversal.traverse(compiler, root, this);
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/Es2020RewriteDynamicImport.java:20]: This should follow the pattern usual to transplation passes.

1. Use `TranspilationPasses.processTranspile()` instead of `NodeTraversal.traverse()`
2. Use `TranspilationPasses.maybeMarkFeaturesAsTranspiledAway()`

See `Es6ForOfConverter#process()` as an example. (Note `Es6RewriteModules` isn't a good pattern to copy, though it may seem tempting since it's related to modules.)

File: src/com/google/javascript/jscomp/DefaultPassConfig.java (hunk 4 above)
Comment (brad4d) [reply src/com/google/javascript/jscomp/DefaultPassConfig.java:267]: Sorry, scratch that.

It's important that the dynamic import still be in the AST when type checking runs so we have a chance to do the type checking we want there.

The pass should get added at the start of `TranspilationPasses.addPostCheckTranspilationPasses()`.

**Review by brad4d — CHANGES_REQUESTED** [review]

File: src/com/google/javascript/jscomp/AstAnalyzer.java (hunk 3 above)
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/AstAnalyzer.java:528]: you want a `break;` even though there is a `return`?

File: src/com/google/javascript/jscomp/DefaultPassConfig.java (hunk 4 above)
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/DefaultPassConfig.java]: Moved

File: src/com/google/javascript/jscomp/Es2020RewriteDynamicImport.java (hunk 6 above)
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/Es2020RewriteDynamicImport.java]: Done

**Comment by ChadKillingsworth** [comment]
The function node conversion is a great idea - but I'll save that until we can actually make this useful.

File: src/com/google/javascript/jscomp/GatherModuleMetadata.java [hunk 7]
```diff
@@ -213,6 +213,8 @@ public boolean shouldTraverse(NodeTraversal t, Node n, Node parent) {
         case MODULE_BODY:
           currentModule.hasModuleBody = true;
           break;
+        case DYNAMIC_IMPORT:
+          visitDynamicImport(t, n);
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/GatherModuleMetadata.java:217]: Actually, this is where the break is missing.

File: src/com/google/javascript/jscomp/AstAnalyzer.java (hunk 3 above)
Comment (brad4d) [reply src/com/google/javascript/jscomp/AstAnalyzer.java:528]: Sorry, I flagged the wrong spot.

**Review by brad4d — CHANGES_REQUESTED** [review]

File: src/com/google/javascript/jscomp/GatherModuleMetadata.java (hunk 7 above)
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/GatherModuleMetadata.java:217]: That makes a lot more sense - fixed!

**Comment by brad4d** [comment]
I'm doing another trial import...

File: src/com/google/javascript/jscomp/parsing/parser/Parser.java [hunk 8]
```diff
@@ -1112,6 +1112,8 @@ private boolean peekStatementStandard() {
       case TILDE:
       case BANG:
         return true;
+      case IMPORT:
+        return peekType(1) == TokenType.OPEN_PAREN;
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/parsing/parser/Parser.java:1116]: Why is this needed?

I presume we forgot some case that should be added to ParserTest in order to test this line.
Could you add a test case there?

Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/parsing/parser/Parser.java:1116]: I was being overly cautious. I can indeed add a test case and ensure this isn't needed.

File: src/com/google/javascript/jscomp/DefaultPassConfig.java [hunk 9]
```diff
@@ -64,6 +64,7 @@
 import com.google.javascript.jscomp.modules.ModuleMapCreator;
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/DefaultPassConfig.java:64]: please drop the now-no-op changes to this file

File: src/com/google/javascript/jscomp/TypeCheck.java [hunk 10]
```diff
@@ -944,6 +944,7 @@ public void visit(NodeTraversal t, Node n, Node parent) {
       case IMPORT_SPEC:
       case IMPORT_SPECS:
       case IMPORT_STAR:
+      case DYNAMIC_IMPORT:
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/TypeCheck.java:947]: This isn't right. `import(something)` is typeable.
Its type should have been already set by TypeInference.

DYNAMIC_IMPORT needs its own case that calls a dedicated `visitDynamicImport()` method.
That method should assert that the child node is typed as string and that the `DYNAMIC_IMPORT` node itself has already been given a type by TypeInference.

If you look around at the other `visitBlah()` methods, I'm sure you can find examples of how to do this.

Comment (brad4d) [reply src/com/google/javascript/jscomp/TypeCheck.java:947]: Also, need to add test cases to TypeCheckNoTranspileTest and TypeInferenceTest.

File: src/com/google/javascript/jscomp/GatherModuleMetadata.java [hunk 11]
```diff
@@ -213,6 +213,9 @@ public boolean shouldTraverse(NodeTraversal t, Node n, Node parent) {
         case MODULE_BODY:
           currentModule.hasModuleBody = true;
           break;
+        case DYNAMIC_IMPORT:
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/GatherModuleMetadata.java:216]: Could you add a test case to cover this to `GatherModuleMetadataTest`?

Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/GatherModuleMetadata.java:216]: Done

File: src/com/google/javascript/jscomp/TypeCheck.java (hunk 10 above)
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/TypeCheck.java]: That helped - I refactored this significantly. I added a check so that the import specifier cannot be undefined or null.

File: src/com/google/javascript/jscomp/DefaultPassConfig.java (hunk 9 above)
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/DefaultPassConfig.java]: done

File: src/com/google/javascript/jscomp/parsing/parser/Parser.java (hunk 8 above)
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/parsing/parser/Parser.java]: It wasn't needed. I removed it and added a test case to prove it wasn't needed.

**Comment by ChadKillingsworth** [comment]
@brad4d Spent a lot of time this weekend tracing through code and was able to finally figure out how to get the specific type of a dynamic import expression.

**Comment by brad4d** [comment]
Importing latest for further review/testiing...

**Review by brad4d — CHANGES_REQUESTED** [review]
I think this is almost ready to land.

My new comments are mostly requests for tests.

General rule we try to follow: If you modify Foo, there should be some corresponding update to FooTest.

Very nice that you worked out where to get the type information for the dynamic import nodes.
That code looks solid to me.

I do need to ask for one other thing, though.
Since this behavior is definitely on the experimental side, I need to make sure people don't trip over it.

Could you add a CommandLineRunner flag and corresponding CompilerOptions field that must be enabled in order for the pass-through behavior to occur? When not enabled, the compiler should report an error as it does now.

Thanks,
Bradford

File: src/com/google/javascript/jscomp/GatherModuleMetadata.java [hunk 12]
```diff
@@ -233,6 +236,15 @@ private void visitImportOrExport(NodeTraversal t, Node importOrExport) {
       }
     }
 
+    private void visitDynamicImport(NodeTraversal t, Node dynamicImport) {
```
[the remark this answers is not in the document]
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/GatherModuleMetadata.java]: done

File: src/com/google/javascript/jscomp/AstValidator.java [hunk 13]
```diff
@@ -433,6 +433,11 @@ public void validateExpression(Node n) {
         validateAwait(n);
         return;
 
+      case DYNAMIC_IMPORT:
```
[the remark this answers is not in the document]
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/AstValidator.java:436]: done

File: src/com/google/javascript/jscomp/PureFunctionIdentifier.java [hunk 14]
```diff
@@ -879,6 +879,12 @@ private void updateSideEffectsForNode(
           }
           break;
 
+        case DYNAMIC_IMPORT:
```
[the remark this answers is not in the document]
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/PureFunctionIdentifier.java:882]: done

File: test/com/google/javascript/jscomp/TypeInferenceTest.java [hunk 15]
```diff
@@ -3751,6 +3765,31 @@ public void testShneTightensUnknownOperandOnRhs() {
...
+            registry.getNativeObjectType(JSTypeNative.PROMISE_TYPE),
+            registry.getNativeType(JSTypeNative.UNKNOWN_TYPE));
+    assertType(getType("foo")).isSubtypeOf(promiseOfUnknownType);
+  }
+
+  @Test
+  public void testDynamicImport2() {
+    withModules(ImmutableList.<String>builder()
+        .add("export default 1; export /** @return {string} */ function Bar() { return 'bar'; };")
+        .add("const foo = import('./testcode0');")
```
[the remark this answers is not in the document]
Comment (ChadKillingsworth) [reply test/com/google/javascript/jscomp/TypeInferenceTest.java]: done

File: test/com/google/javascript/jscomp/TypeInferenceTest.java [hunk 16]
```diff
@@ -3751,6 +3765,31 @@ public void testShneTightensUnknownOperandOnRhs() {
...
+    JSType promiseOfUnknownType =
+        registry.createTemplatizedType(
+            registry.getNativeObjectType(JSTypeNative.PROMISE_TYPE),
+            registry.getNativeType(JSTypeNative.UNKNOWN_TYPE));
+    assertType(getType("foo")).isSubtypeOf(promiseOfUnknownType);
+  }
+
+  @Test
+  public void testDynamicImport2() {
+    withModules(ImmutableList.<String>builder()
```
[the remark this answers is not in the document]
Comment (ChadKillingsworth) [reply test/com/google/javascript/jscomp/TypeInferenceTest.java]: done

File: test/com/google/javascript/jscomp/RewriteDynamicImportTest.java [hunk 17]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2014 The Closure Compiler Authors.
```
[the remark this answers is not in the document]
Comment (ChadKillingsworth) [reply test/com/google/javascript/jscomp/RewriteDynamicImportTest.java]: done

File: src/com/google/javascript/jscomp/TypeValidator.java [hunk 18]
```diff
@@ -345,6 +350,21 @@ void expectAsyncGeneratorSupertype(Node n, JSType type, String msg) {
     }
   }
 
+  /**
+   * Expect the type to be a supertype of `Promise`.
+   *
+   * <p>`Promise` is the <em>lower</em> bound of the declared return type, since that's what async
+   * functions always return; the user can't return an instance of a more specific type.
+   */
+  void expectValidDynamicImportType(Node n, JSType type) {
```
[the remark this answers is not in the document]
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/TypeValidator.java]: This entire function ended up being unnecessary

File: src/com/google/javascript/jscomp/TypeValidator.java [hunk 19]
```diff
@@ -345,6 +350,21 @@ void expectAsyncGeneratorSupertype(Node n, JSType type, String msg) {
...
+    if (promiseOfUnknownType.isSubtypeOf(type)) {
```
[the remark this answers is not in the document]
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/TypeValidator.java]: Trying to write a test for this made it very clear this was impossible to trigger. So I just removed the the entire function.

File: src/com/google/javascript/jscomp/AstAnalyzer.java [hunk 20]
```diff
@@ -261,6 +261,10 @@ private boolean checkForStateChangeHelper(Node n, boolean checkForNewObjects) {
         // Variable declarations are side-effects.
         return true;
 
+      // import() expressions have side effects
```
[the remark this answers is not in the document]
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/AstAnalyzer.java:264]: done

File: src/com/google/javascript/jscomp/NodeUtil.java [hunk 21]
```diff
@@ -4754,6 +4754,7 @@ static boolean evaluatesToLocalValue(Node value) {
             && evaluatesToLocalValue(value.getLastChild());
       case THIS:
       case SUPER:
+      case DYNAMIC_IMPORT:
```
[the remark this answers is not in the document]
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/NodeUtil.java:4757]: done

File: src/com/google/javascript/jscomp/TypeValidator.java [hunk 22]
```diff
@@ -345,6 +350,21 @@ void expectAsyncGeneratorSupertype(Node n, JSType type, String msg) {
...
+      return;
+    }
+
+    JSError err = JSError.make(n, INVALID_DYNAMIC_IMPORT_TYPE, type.toString());
+    registerMismatchAndReport(type, promiseOfUnknownType, err);
```
[the remark this answers is not in the document]
Comment (ChadKillingsworth) [reply src/com/google/javascript/jscomp/TypeValidator.java]: I added additional tests to TypeCheckTest and removed this entire function. I found it impossible to trigger this case. One note: this did outline a parse exception when trying to type-case a dynamic import. I fixed that too.

**Comment by ChadKillingsworth** [comment]
I've now added a `--allow_dynamic_import` flag. Without this flag a special check pass will error if encountering any DYNAMIC_IMPORT node.

**Comment by brad4d** [comment]
I've imported the latest version and started standard testing on it.

**Comment by brad4d** [comment]
I've sent this to @concavelenz for internal review

**Comment by brad4d** [comment]
Internal review is nearly finished. Addressed more review comments from @concavelenz today.
